ext/rubymain.cpp

```diff
@@ -744,24 +744,33 @@ static VALUE t_set_sock_opt (VALUE self UNUSED, VALUE signature, VALUE lev, VALU
 	const void *v;
 	socklen_t len;
 
-	switch (TYPE(optval)) {
-	case T_FIXNUM:
-		i = FIX2INT(optval);
-		goto numval;
-	case T_FALSE:
-		i = 0;
-		goto numval;
-	case T_TRUE:
-		i = 1;
-		numval:
-		v = (void*)&i; len = sizeof(i);
-		break;
-	default:
-		StringValue(optval);
-		v = RSTRING_PTR(optval);
-		len = RSTRING_LENINT(optval);
-		break;
-	}
+  if (option == SO_BINDTODEVICE) {
+    struct ifreq interface; // create struct
+    StringValue(optval); // parse the ruby string
+
+    strncpy(interface.ifr_name, RSTRING_PTR(optval), IFNAMSIZ); // copy interface to struct
```

Comment:
This should be `StringValueCStr(optval)` to ensure that it is nul-terminated. The RSTRING_PTR can point to a Ruby value that doesn't have a nul-terminator; you need RSTRING_LEN to read only the number of valid bytes. And since `strncpy` does not guarantee a nul-terminator, if the Ruby string doesn't provide one then you may get quite a bit of garbage in there!